### PR TITLE
Fixed CS4014  - Not awaiting a Task

### PR DESCRIPTION
**Pull request by NimaAra**
Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.

### CODE REVIEW

**Comment by vtortola** [comment]
Hi, that un-awaited task is intentional. HTTP negotiation happens asynchronously. Basically, there is a task awaiting sockets, and each time one connects, it calls the negotiation in another task and gets back to listen for more connecting sockets. If you await it, the component won't listen for more incoming connections until the negotiation has finished, and we do not want that.

Cheers.

**Comment by NimaAra** [comment]
Hi, 

But you are already calling `WorkAsync()` asynchronously from line [50](https://github.com/vtortola/WebSocketListener/blob/master/vtortola.WebSockets/Http/HttpNegotiationQueue.cs#L50) using `Task.Run(WorkAsync)`

I actually think a more efficient approach would be to have a Producer/Consumer queue with a set number of worker threads (think long-running tasks) assigned to it to process the offloaded negotiation work instead of sending them all as Tasks to the ThreadPool. I even think assigning more than 1 worker thread for the negotiation is unnecessary but of course your mileage may vary and profiling can determine the best configuration. 

In summary I usually prefer to have more control when it comes to creating and offloading (_possibly_) too many tasks over to the ThreadPool. This helps reduce pressure on GC and the ThreadPool. 

I have been bitten by this in the past! ;-)

**Comment by vtortola** [comment]
Yes. The task running `WorkAsync` is launched once, and its sole mission is to get incoming sockets and launch negotiations in parallel, that is why the task running `NegotiateWebSocket` is called asyncrhonously, meaning that multiple sockets can be negotiated at the same time independently, and also keep accepting sockets at the same time. Your pull-request will block `WorkAsync` labour until negotiation is completed, causing that only one HTTP negotiation can be made at the time, an not sockets are being accepted while that negotiation is being performed. There is a configurable limitation on the number of HTTP negotiations that can be in parallel, so it is not big deal.

Anyway, if you put together a demo where another approach works better I will be definitely interested :)

**Comment by NimaAra** [comment]
Okay, I am currently trying to integrate your work on another project I am building ([Easy.Transport](https://github.com/NimaAra/Easy.Transport)) currently it is using my fork of [WebSocket-Sharp](https://github.com/sta/websocket-sharp) which has been tuned to be async and more efficient than the original project. Although I have not committed the final code yet.

Once I have finished integrating your project I can have a good idea of performance comparing the two.

**Comment by vtortola** [comment]
It sounds good to me :+1:
